src/main.rs

```diff
@@ -0,0 +1,270 @@
+        };
+
+        return Render::render(res, "public/index.tpl", &data)
+    });
+
+
+    let _ = server.listen("127.0.0.1:6767").unwrap();
+}
+
+fn validate_password(pass: &str) -> Result<(), String> {
+    if pass.len() < 10 {
+        return Err("password must be at least 10 characters long".to_string());
+    }
+    Ok(())
+}
+
+fn validate_uname_and_token(token_dir: &str, uname: &str, token: &str) -> Result<(), String> {
+    // token database is just the filesystem (fuckit shipit).
+    // Tokens are stored in the hierarchy "tokens/$token" relative to the program's cwd.
+    // The token file contains the string "username".
+    //
+    // For obvious security reasons, '.' and '/' should be invalid in the token. Just assert it's
+    // alphanumeric for simplicity, which solves that.
+    if !token.chars().all(|c| c.is_ascii() && c.is_alphanumeric()) {
+        return Err("token must be ascii/alphanumeric".to_string());
+    }
+
+    let token_path = Path::new(token_dir).join(format!("tokens/{}", token).as_str());
+    let mut f = match File::open(token_path) {
+        Ok(f) => f,
+        Err(_) => return Err("invalid token".to_string()), // TODO log non-ENOENT errs
+    };
+
+    let mut token_uname = String::new();
+    if !f.read_to_string(&mut token_uname).is_ok() {
+        return Err("invalid token + username".to_string());
+    }
+
+    if token_uname.trim() != uname {
+        return Err("invalid token + username".to_string());
+    }
+    Ok(())
+}
+
+// delete_token should be called after validate_uname_and_token since it assumes the token has been
+// validated
+fn delete_token(token_dir: &str, token: &str) -> std::io::Result<()> {
+    let token_path = Path::new(token_dir).join(format!("tokens/{}", token).as_str());
+    std::fs::remove_file(token_path)
+}
+
+
+
+fn set_new_password(db_conn: &str, uname: &str, password_hash: &str) -> Result<(), SetPwError> {
+    // TODO, connection pooling a level above this function
+    let conn = Connection::connect(db_conn, TlsMode::None)?;
+    // Based on the synapse readme here: https://github.com/matrix-org/synapse/blob/f9834a3d1a25d0a715718a53e10752399985e3aa/README.rst#password-reset
+    // UPDATE users SET password_hash='$2a$12$xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' WHERE
+    // name='@test:test.com';
+    // is the query we want
+    let updates = conn.execute("UPDATE users SET password_hash = $1 WHERE name = $2",
+                 &[&password_hash, &uname])?;
+
+    match updates {
+        0 => Err(SetPwError::InvalidUserError()),
+        1 => Ok(()),
+        _ => Err(SetPwError::UnexpectedError()),
+    }
+}
+
+fn hash_password(password: &str, pepper: &str) -> String {
+    // This function closely mimics
+    // https://github.com/matrix-org/synapse/blob/9bba6ebaa903a81cd94fada114aa71e20b685adb/scripts/hash_password
+
+    // Match the default from synapse arbitrarily
+    let bcrypt_rounds = 12;
+
+    let mut salt = [0u8; 16];
+    let mut rng = OsRng::new().unwrap();
```

Comment:
This should never actually happen since I think OsRng will be Ok on all platforms Rust currently supports, but a panic here can possibly bring down the service. Would be safer for hash_password to return a Result.